File: 2024/day18.go
package main

import (
	"sort"

	"lukechampine.com/advent/utils"
)

var input = utils.Input(2024, 18)

func path(bytes []utils.Pos) int {
	corrupt := make(map[utils.Pos]bool)
	for _, p := range bytes {
		corrupt[p] = true
	}
	g := utils.Grid{X: 70, Y: 70}
	queue := []utils.Pos{{X: 0, Y: 0}}
	seen := make(map[utils.Pos]int)
	best := 999999
	for len(queue) > 0 {
		p := queue[0]
		queue = queue[1:]
		if p == (utils.Pos{X: g.X, Y: g.Y}) {
			best = utils.Min(best, seen[p])
			continue
		}
		for _, m := range p.ValidMoves(g) {
			if s, ok := seen[m]; (!ok || s > seen[p]+1) && !corrupt[m] {
				queue = append(queue, m)
				seen[m] = seen[p] + 1
			}
		}
	}
	return best
}

func main() {
	var bytes []utils.Pos
	ints := utils.ExtractInts(input)
	for i := 0; i < len(ints); i += 2 {
		bytes = append(bytes, utils.Pos{X: ints[i], Y: ints[i+1]})
	}
	utils.Println(path(bytes[:1024]))

	n := sort.Search(len(bytes), func(n int) bool {
		return path(bytes[:n]) == 999999
	})
	utils.Print(bytes[n-1].X, ",", bytes[n-1].Y, "\n", n)
}

File: 2024/rust/src/day18.rs
use crate::utils;
use std::collections::HashMap;

fn parse(input: &str) -> Vec<utils::Point> {
    utils::ints(input)
        .collect::<Vec<i64>>()
        .chunks(2)
        .map(|w| (w[0], w[1]).into())
        .collect()
}

fn best_path(bytes: Vec<utils::Point>) -> Option<i64> {
    let mut grid = utils::Grid::new(71, 71, b'.');
    bytes.iter().for_each(|&p| grid[p] = b'#');

    let mut best: Option<i64> = None;
    let mut dists = HashMap::new();
    let mut queue = vec![((0, 0).into(), 0)];
    while let Some((p, d)) = queue.pop() {
        if p == (70, 70).into() {
            best = Some(best.map_or(d, |b| b.min(d)));
            continue;
        }
        grid.adj(p).for_each(|(m, c)| {
            let e = dists.entry(m).or_insert(i64::MAX);
            if c != b'#' && d + 1 < *e {
                *e = d + 1;
                queue.push((m, d + 1));
            }
        });
    }
    best
}

pub fn part1(input: &str) -> String {
    best_path(parse(input)[..1024].into()).unwrap().to_string()
}

pub fn part2(input: &str) -> String {
    let bytes = parse(input);
    let n = (0..bytes.len())
        .collect::<Vec<_>>()
        .partition_point(|&n| best_path(bytes[0..n].into()).is_some());
    format!("{},{}", bytes[n - 1].x, bytes[n - 1].y)
}
